### PR TITLE
BACK-1370 Improved operations fetch

### PR DESCRIPTION
**Pull request by mbasire-ledger**
## What is this about?

Intend to reduce the number of database requests when fetching operations.
The wrapped Transaction inside an operation is no more an option which is more meaningful.

Before fetching N operations, there were the following requests:
 (1 req for N operations fetching) + N * ( 1 req for tx + 1 inputs req + 1 outputs req)

It becomes 
(1 req for N operation fetching) + N / (hardcoded number)  * ( 1 req for inputs + 1 for for outputs ) 

## Cute picture of lama
lama

### CODE REVIEW

**Comment by mbasire-ledger** [comment]
I expect to have at least 1 Input and 1 Ouput for each Transaction. Is it right ?

**Comment by mbasire-ledger** [comment]
This PR is waiting to another PR whare the unconfirmed transactions on stored in the same tables than the confirmed transactions

File: coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueriesIT.scala [hunk 1]
```diff
@@ -49,26 +54,38 @@ class QueriesIT extends AnyFlatSpecLike with Matchers with TestResources {
...
+          opWithTx shouldBe None
 
-          tx.id shouldBe transactionToInsert.id
-          tx.hash shouldBe transactionToInsert.hash
+          val inputs: Seq[InputView] = inputsWithOutputs._1
 
-          tx.inputs should have size 1
-          tx.inputs.head.value shouldBe 80000
+          inputs should have size 1
+          inputs.head.value shouldBe 80000
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueriesIT.scala:83]: We could also check that the hash is consistent (to be added in InputView ?)

File: coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueryUtils.scala [hunk 2]
```diff
@@ -1,23 +1,45 @@
...
 
-object QueryUtils {
+import java.util.UUID
 
-  def fetchTx(db: Transactor[IO], accountId: UUID, hash: String): IO[Option[TransactionView]] = {
+object QueryUtils {
+  def fetchInputAndOutputs(
+      db: Transactor[IO],
+      accountId: UUID,
+      hash: String
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueryUtils.scala:25]: to rename txHash

File: coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueryUtils.scala [hunk 3]
```diff
@@ -1,23 +1,45 @@
...
+  ): IO[(List[InputView], List[OutputView])] = {
     OperationQueries
-      .fetchTransaction(accountId, hash)
+      .fetchInputsWithOutputsOrderedByTxHash(accountId, Sort.Descending, NonEmptyList.one(hash))
       .transact(db)
+      .map(_._2)
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueryUtils.scala:30]: using the name of the parameter here would be clearer

File: coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueryUtils.scala [hunk 4]
```diff
@@ -1,23 +1,45 @@
...
+      .compile
+      .toList
+      .map(_.head)
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueryUtils.scala:33]:
```suggestion
      .last
```

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/models/implicits.scala [hunk 5]
```diff
@@ -65,13 +65,15 @@ object implicits {
 
   // Needs implicit because of the Transaction "None"
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/models/implicits.scala:66]: Invalid comment

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala [hunk 6]
```diff
@@ -14,31 +13,59 @@ import doobie._
...
+      fees: BigInt,
+      time: Instant,
+      blockHeight: Option[Long]
+  )
+
+  def fetchInputsWithOutputsOrderedByTxHash(
+      accountId: UUID,
+      sort: Sort,
+      txHashes: NonEmptyList[String]
+  ): Stream[doobie.ConnectionIO, (String, (List[InputView], List[OutputView]))] = {
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:48]: Having a `String` type for the txhashes is not strong enough and we lose context

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala [hunk 7]
```diff
@@ -164,81 +191,96 @@ object OperationQueries extends IOLogging {
...
-      .query[InputView]
-      .stream
+      sort: Sort,
+      txHashes: NonEmptyList[String]
+  ) = {
+
+    val belongsToTxs = allTxHashes(txHashes)
+
+    (sql"""
+          SELECT i.hash, i.output_hash, i.output_index, i.input_index, i.value, i.address, i.script_signature, i.txinwitness, i.sequence, i.derivation
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:206]: should select `o.hash`: if there's a line matching the WHERE but without input, we'll select `NULL` here

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala [hunk 8]
```diff
@@ -164,81 +191,96 @@ object OperationQueries extends IOLogging {
...
-      .stream
+      sort: Sort,
+      txHashes: NonEmptyList[String]
+  ) = {
+
+    val belongsToTxs = allTxHashes(txHashes)
+
+    (
+      sql"""
+          SELECT output.hash, output.output_index, output.value, output.address, output.script_hex, output.change_type, output.derivation
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:225]: should select o.hash: if there's a line matching the WHERE but without input, we'll select NULL here

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala [hunk 9]
```diff
@@ -164,81 +191,96 @@ object OperationQueries extends IOLogging {
...
+      txHashes: NonEmptyList[String]
+  ) = {
+
+    val belongsToTxs = allTxHashes(txHashes)
+
+    (sql"""
+          SELECT i.hash, i.output_hash, i.output_index, i.input_index, i.value, i.address, i.script_signature, i.txinwitness, i.sequence, i.derivation
+            FROM operation o 
+            LEFT JOIN input i on i.account_id = o.account_id and i.hash = o.hash
+           WHERE i.account_id = $accountId
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:209]: should also filter on o.account_id

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala [hunk 10]
```diff
@@ -164,81 +191,96 @@ object OperationQueries extends IOLogging {
...
+  ) = {
+
+    val belongsToTxs = allTxHashes(txHashes)
+
+    (
+      sql"""
+          SELECT output.hash, output.output_index, output.value, output.address, output.script_hex, output.change_type, output.derivation
+            FROM operation o  
+            LEFT JOIN output on output.account_id = o.account_id and output.hash = o.hash
+           WHERE output.account_id = $accountId
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:228]: should also filter on o.account_id

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala [hunk 11]
```diff
@@ -164,81 +191,96 @@ object OperationQueries extends IOLogging {
...
+                OR o.block_height IS NULL)
+         """ ++ operationOrder(sort) ++ limitF ++ offsetF
+
+    (operationWithTx ++ filter)
+      .query[(Op, Tx)]
+      .stream
   }
 
   def findOperation(
       accountId: Operation.AccountId,
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:271]: accountId is not necessary, but maybe for a later cleanup

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala [hunk 12]
```diff
@@ -34,13 +27,32 @@ class OperationService(
...
+            .flatMap(txHashes =>
+              OperationQueries
+                .fetchInputsWithOutputsOrderedByTxHash(accountId, sort, txHashes)
+            )
+
+          Stream
+            .chunk(ops)
+            .covary[ConnectionIO]
+            .zip(inputsAndOutputs)
+            .flatMap { case ((txHash, operationsSentAndReceive), (txHash1, (inputs, outputs))) =>
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala:46]:
```suggestion
            .flatMap { case ((txHash, operationsSentAndReceived), (txHash1, (inputs, outputs))) =>
```

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala [hunk 13]
```diff
@@ -34,13 +27,32 @@ class OperationService(
...
+              Stream
+                .chunk(operationsSentAndReceive)
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala:48]:
```suggestion
                .chunk(operationsSentAndReceived)
```

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala [hunk 14]
```diff
@@ -34,13 +27,32 @@ class OperationService(
...
       opsWithTx <- OperationQueries
         .fetchOperations(accountId, blockHeight, sort, Some(limit + 1), Some(offset))
-        .transact(db)
-        .parEvalMap(maxConcurrent) { op =>
-          OperationQueries
-            .fetchTransaction(op.accountId, op.hash)
-            .transact(db)
-            .map(tx => op.copy(transaction = tx))
+        .groupAdjacentBy(_._1.hash) // many operations by hash (RECEIVED AND SENT)
+        .chunkN(5)                  // we have to figure out which value is more suitable here
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala:31]: 5 could be a named variable instead

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala [hunk 15]
```diff
@@ -179,3 +230,12 @@ class OperationService(
         .flatMap(Stream.chunk)
 
 }
+
+object OperationService {
+
+  sealed trait Error
+  case class NoMatchingTransactionError(operationId: Operation.UID)
+      extends Throwable(s"No transaction found for $operationId")
+      with Error
```
Comment (gagbo) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationService.scala:239]: unused at this time

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala (hunk 11 above)
Comment (MiguelBasire) [reply coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:271]: You are right Operation.UID is f(account_id, ...). I check if there is a particular reason to still use account_id.

Comment (MiguelBasire) [reply coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:271]: account_id is not even indexed in the database anymore ! 
I guess it was deleted when Operation.UID has been added. account_id and hash are referenced by all tables associated with operation.

Comment (MiguelBasire) [reply coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/services/OperationQueries.scala:271]: Ok, there is an `operation_compute_balance` index with account_id . This is not too bad.

**Review by nukky94 — APPROVED** [review]

**Review by estrauser-ledger — APPROVED** [review]

**Comment by estrauser-ledger** [comment]
We talked about using case class instead of (txHash, (inputList, outputList))

**Pull request merged**